File: app/src/main/java/com/example/mvctutorial/designpattern/_13chain_of_resposibility/part1/Calculator.java
package com.example.mvctutorial.designpattern._13chain_of_resposibility.part1;

public abstract class Calculator {
    Calculator nextCalculator;

    public void setNextCalculator(Calculator nextCalculator) {
        this.nextCalculator = nextCalculator;
    }

    boolean process(Request request) {

        if (nextCalculator == null) return false;

        if (operator(request)){
            return true;
        }else{
            return nextCalculator.operator(request);
        }

    }

    abstract protected boolean operator(Request request);
}

File: app/src/main/java/com/example/mvctutorial/designpattern/_13chain_of_resposibility/part1/Calculator.kt
package com.example.mvctutorial.designpattern._13chain_of_resposibility.part1

abstract class Calculator {

    var nextCalculator: Calculator? = null

    fun process(request: Request): Boolean {

        nextCalculator ?: return false

        return if (operator(request)) true
        else nextCalculator!!.operator(request)
    }

    protected abstract fun operator(request: Request): Boolean
}